Compiler Crashes Attempting To Compile ZIO Flow On 3.0.0-RC1

The issue can be seen in this commit - https://github.com/zio/zio-flow/tree/a28bd7cfb777bd2a8d50c80a80b4686e1d1d9fd2

I've not had a chance to create a minimal repro yet. I'll try to do that and add it later.

The error message is unfortunately very long.

```
[error] ## Exception when compiling 23 sources to /Users/aiswaryaprakasan/Projects/zio/zio-flow/zio-flow/jvm/target/scala-3.0.0-RC1/classes
[error] java.lang.AssertionError: NoDenotation.owner
[error] dotty.tools.dotc.core.SymDenotations$NoDenotation$.owner(SymDenotations.scala:2394)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$2(TreePickler.scala:422)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:423)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$13(TreePickler.scala:506)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:506)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$41$$anonfun$1(TreePickler.scala:502)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$12(TreePickler.scala:502)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:503)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTreeUnlessEmpty(TreePickler.scala:330)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:345)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:566)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:592)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:593)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:342)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:568)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:592)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:593)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:342)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:568)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$29(TreePickler.scala:608)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:608)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:778)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:778)
[error] dotty.tools.dotc.transform.Pickler.run$$anonfun$3$$anonfun$2(Pickler.scala:69)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:106)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.transform.Pickler.run(Pickler.scala:106)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error] scala.collection.immutable.List.map(List.scala:250)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error] dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:111)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:230)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] dotty.tools.dotc.Run.compile(Run.scala:150)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2177)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2134)
[error] sbt.internal.io.Retry$.apply(Retry.scala:40)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2130)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
[error] java.base/java.lang.Thread.run(Thread.java:832)
[error]            
[error] stack trace is suppressed; run last zioFlowJVM / Compile / compileIncremental for the full output
[error] (zioFlowJVM / Compile / compileIncremental) java.lang.AssertionError: NoDenotation.owner
```

Minimized:

```scala
sealed trait Expr[+A] extends ExprTuple[A]

case class Lit[A](value: A) extends Expr[A]

trait ExprTuple[+A] {
  def _1(implicit ev: ExprTuple.Has2[A]): Expr[ev._1] = ???
  def _2(implicit ev: ExprTuple.Has2[A]): Expr[ev._2] = ???
}

object ExprTuple {
  trait Has2[-A] { type _1; type _2 }
  object Has2 {
    type Aux[-A, A1, A2] = Has2[A] { type _1 = A1; type _2 = A2 }
  }

  implicit def tuple2[A, B]: Has2.Aux[(A, B), A, B] = ???
}
```

<details>
  <summary>Compiler Output (3.0.0-RC1)</summary>

```scala
error when pickling tree Lit.this._1
error when pickling tree case 0 => Lit.this._1
error when pickling tree n match 
  {
    case 0 => Lit.this._1
    case _ => throw new IndexOutOfBoundsException(n.toString())
  }
error when pickling tree override def productElement(n: Int): Any = 
  n match 
    {
      case 0 => this._1
      case _ => throw new IndexOutOfBoundsException(n.toString())
    }
error when pickling tree [A >: Nothing <: Any](value: A) extends Object(), Expr[Lit.this.A], _root_.scala
  .
Product, _root_.scala.Serializable {
  override def hashCode(): Int = scala.runtime.ScalaRunTime._hashCode(this)
  override def equals(x$0: Any): Boolean = 
    this.eq(x$0.$asInstanceOf$[Object]).||(
      x$0 match 
        {
          case x$0 @ _:Lit[A] @unchecked => 
            this.value.==(x$0.value).&&(x$0.canEqual(this))
          case _ => false
        }
    )
  override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
  override def canEqual(that: Any): Boolean = 
    that.isInstanceOf[Lit[A] @unchecked]
  override def productArity: Int = 1
  override def productPrefix: String = "Lit"
  override def productElement(n: Int): Any = 
    n match 
      {
        case 0 => this._1
        case _ => throw new IndexOutOfBoundsException(n.toString())
      }
  override def productElementName(n: Int): String = 
    n match 
      {
        case 0 => "value"
        case _ => throw new IndexOutOfBoundsException(n.toString())
      }
  A
  val value: A
  def copy[A](value: A): Lit[A] = new Lit[A](value)
  def copy$default$1[A]: A = 
    Lit.this.value:(Lit.this.value : A) @uncheckedVariance
  def _1: A = this.value
}
error when pickling tree @scala.annotation.internal.SourceFile("Expr.scala") case class Lit[
  A >: Nothing <: Any
](value: A) extends Object(), Expr[Lit.this.A], _root_.scala.Product, _root_.
  scala
.Serializable {
  override def hashCode(): Int = scala.runtime.ScalaRunTime._hashCode(this)
  override def equals(x$0: Any): Boolean = 
    this.eq(x$0.$asInstanceOf$[Object]).||(
      x$0 match 
        {
          case x$0 @ _:Lit[A] @unchecked => 
            this.value.==(x$0.value).&&(x$0.canEqual(this))
          case _ => false
        }
    )
  override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
  override def canEqual(that: Any): Boolean = 
    that.isInstanceOf[Lit[A] @unchecked]
  override def productArity: Int = 1
  override def productPrefix: String = "Lit"
  override def productElement(n: Int): Any = 
    n match 
      {
        case 0 => this._1
        case _ => throw new IndexOutOfBoundsException(n.toString())
      }
  override def productElementName(n: Int): String = 
    n match 
      {
        case 0 => "value"
        case _ => throw new IndexOutOfBoundsException(n.toString())
      }
  A
  val value: A
  def copy[A](value: A): Lit[A] = new Lit[A](value)
  def copy$default$1[A]: A = 
    Lit.this.value:(Lit.this.value : A) @uncheckedVariance
  def _1: A = this.value
}
error when pickling tree package <empty> {
  @scala.annotation.internal.SourceFile("Expr.scala") case class Lit[
    A >: Nothing <: Any
  ](value: A) extends Object(), Expr[Lit.this.A], _root_.scala.Product, _root_.
    scala
  .Serializable {
    override def hashCode(): Int = scala.runtime.ScalaRunTime._hashCode(this)
    override def equals(x$0: Any): Boolean = 
      this.eq(x$0.$asInstanceOf$[Object]).||(
        x$0 match 
          {
            case x$0 @ _:Lit[A] @unchecked => 
              this.value.==(x$0.value).&&(x$0.canEqual(this))
            case _ => false
          }
      )
    override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
    override def canEqual(that: Any): Boolean = 
      that.isInstanceOf[Lit[A] @unchecked]
    override def productArity: Int = 1
    override def productPrefix: String = "Lit"
    override def productElement(n: Int): Any = 
      n match 
        {
          case 0 => this._1
          case _ => throw new IndexOutOfBoundsException(n.toString())
        }
    override def productElementName(n: Int): String = 
      n match 
        {
          case 0 => "value"
          case _ => throw new IndexOutOfBoundsException(n.toString())
        }
    A
    val value: A
    def copy[A](value: A): Lit[A] = new Lit[A](value)
    def copy$default$1[A]: A = 
      Lit.this.value:(Lit.this.value : A) @uncheckedVariance
    def _1: A = this.value
  }
  final lazy module val Lit: Lit$ = new Lit$()
  @scala.annotation.internal.SourceFile("Expr.scala") final module class Lit$()
     extends
   AnyRef(), scala.deriving.Mirror.Product { this: Lit.type =>
    private def writeReplace(): AnyRef = 
      new scala.runtime.ModuleSerializationProxy(classOf[Lit.type])
    def apply[A](value: A): Lit[A] = new Lit[A](value)
    def unapply[A](x$1: Lit[A]): Lit[A] = x$1
    override def toString: String = "Lit"
    type MirroredMonoType = Lit[? <: AnyKind]
    def fromProduct(x$0: Product): Lit.MirroredMonoType = 
      new Lit[Any](x$0.productElement(0))
  }
}
exception occurred while compiling Expr.scala
java.lang.AssertionError: NoDenotation.owner while compiling Expr.scala
Exception in thread "main" java.lang.AssertionError: NoDenotation.owner
	at dotty.tools.dotc.core.SymDenotations$NoDenotation$.owner(SymDenotations.scala:2394)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$2(TreePickler.scala:422)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:423)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$13(TreePickler.scala:506)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:506)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$41$$anonfun$1(TreePickler.scala:502)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$12(TreePickler.scala:502)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:503)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTreeUnlessEmpty(TreePickler.scala:330)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:345)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:566)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:592)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:593)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:342)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:568)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$29(TreePickler.scala:608)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:608)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:778)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:778)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$3$$anonfun$2(Pickler.scala:69)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:106)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:106)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:111)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:230)
	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
	at dotty.tools.dotc.Run.compile(Run.scala:150)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
	at dotty.tools.dotc.Driver.process(Driver.scala:186)
	at dotty.tools.dotc.Driver.process(Driver.scala:155)
	at dotty.tools.dotc.Driver.process(Driver.scala:167)
	at dotty.tools.dotc.Driver.main(Driver.scala:194)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala
+++ b/compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala
@@ -174,7 +174,8 @@ class SyntheticMembers(thisPhase: DenotTransformer) {
     def productElementBody(arity: Int, index: Tree)(using Context): Tree = {
       // case N => _${N + 1}
       val cases = 0.until(arity).map { i =>
-        CaseDef(Literal(Constant(i)), EmptyTree, Select(This(clazz), nme.selectorName(i)))
+        val sel = This(clazz).select(nme.selectorName(i), _.info.isParameterless)
+        CaseDef(Literal(Constant(i)), EmptyTree, sel)
       }
 
       Match(index, (cases :+ generateIOBECase(index)).toList)

--- a/tests/pos/i11725.scala
+++ b/tests/pos/i11725.scala
@@ -1,0 +1,17 @@
+sealed trait Expr[+A] extends ExprTuple[A]
+
+case class Lit[A](value: A) extends Expr[A]
+
+trait ExprTuple[+A] {
+  def _1(implicit ev: ExprTuple.Has2[A]): Expr[ev._1] = ???
+  def _2(implicit ev: ExprTuple.Has2[A]): Expr[ev._2] = ???
+}
+
+object ExprTuple {
+  trait Has2[-A] { type _1; type _2 }
+  object Has2 {
+    type Aux[-A, A1, A2] = Has2[A] { type _1 = A1; type _2 = A2 }
+  }
+
+  implicit def tuple2[A, B]: Has2.Aux[(A, B), A, B] = ???
+}

